test/common/network/udp_listener_impl_test.cc

```diff
@@ -301,7 +316,12 @@ TEST_P(UdpListenerImplTest, UdpListenerRecvMsgError) {
   // Inject mocked OsSysCalls implementation to mock a read failure.
   Api::MockOsSysCalls os_sys_calls;
   TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
+#if ENVOY_MMSG_MORE
+  EXPECT_CALL(os_sys_calls, recvmmsg(_, _, _, _, _))
+      .WillOnce(Return(Api::SysCallIntResult{-1, ENOTSUP}));
```

Comment:
Stepping back, WDYT about just having recvmmsg as the only API we expose, and if the OS doesn't support it just implement it using recvmsg? Would that make for a lot less ifdefs?

Reply:
I believe later we will limit number of system calls per event loop. Implementing recvmmsg() with recvmsg() might add a lot of complexity to that.

Reply:
I'm really not thrilled with sprinkling the ENVOY_MMSG_MORE define all through the code. It seems like we should be able to abstract this somewhere? Maybe at the IOHandle layer?

Reply:
Hmm, I would really like to come to some solution here that doesn't sprinkle this macro all over the place. What about a wrapper function that internally either does recvmmsg or recvmsg depending on capability and then returns some number of messages? I think that would spread this around a lot less? @alyssawilk @ggreenway any thoughts here?

Reply:
How about this:

In `OsSysCalls::recvmmsg()`, put the `#ifdef` for whether it is supported, and if it is not supported, return an error code (EUNSPPORTED or something). In the caller, always start with calling recvmmsg on all platforms, and on all platforms have a fallback to `OsSysCalls::recvmsg()` if EUNSUPPORTED is returned. 

In the tests, which use a mock `OsSysCalls`, add coverage for both paths. 

The runtime overhead of calling a virtual function which returns a constant and branching to another path should be minimal. There's only 1 #ifdef then, in OsSysCalls.

Reply:
Alternately we think there's value in the caller knowing if they're signing up for 1 or 16 syscalls where they call recv[m]msg, we could expose a single function for "is this recvmmsg supported" which uses these defines, but allows for clean if(supported) {} else {} across the code and test base.